File: Objetos_2_java/src/main/java/objetos2/java/juegoRol/Personaje.java
package objetos2.java.juegoRol;

public abstract class Personaje {
    String nombre;
    int vida;
    int danioBase;

    public Personaje(String nombre, int vida) {
        this.nombre = nombre;
        this.vida = vida;
        this.danioBase = 10;
    }

    public abstract void atacar(Personaje personaje);

    public void recibirDanio(int cantidad){
        vida -= cantidad;
        if (vida <= 0) vida = 0;
        System.out.println(nombre + " recibe " + cantidad + " de daño. Vida restante: " + vida);
    }

    public boolean estaVivo(){
        return vida > 0;
    }




}

File: Objetos_2_Kotlin/src/main/kotlin/juegoRol/Personaje.kt
package juegoRol

open class Personaje(
    val nombre: String,
    var vida: Int,
    open val danioBase: Int = 10
) {

    open fun atacar(objetivo: Personaje) {

        if (objetivo.vida > 0) {
            print("$nombre ataca a ${objetivo.nombre} con un daño de $danioBase. ")
            objetivo.recibirDanio(danioBase)
        }else{
            println("${objetivo.nombre} ha sido eliminado")
        }
    }

    open fun recibirDanio(danio: Int){
        vida -= danio

        if (vida <= 0){
            vida = 0
        }else{
            println("$nombre tiene $vida de vida restante")
        }
    }
}